gateway-sp-comms/src/management_switch.rs

```diff
@@ -0,0 +1,592 @@
+            &self.known_sps.switches,
+            &self.known_sps.sleds,
+            &self.known_sps.power_controllers,
+        ] {
+            if n < known_sps.len() {
+                return Some(SpSocket {
+                    socket: &self.ports[port.0],
+                    addr: known_sps[n].sp,
+                    port,
+                });
+            }
+            n -= known_sps.len();
+        }
+
+        // We only construct `SwitchPort`s with valid indices, so the only way
+        // to get here is if someone constructs two `ManagementSwitch` instances
+        // (with different port cardinalities) and then mixes up `SwitchPort`s
+        // between them (or other similarly outlandish shenanigans) - fine to
+        // panic in that case.
+        unreachable!("invalid port {:?}", port)
+    }
+}
+
+async fn recv_task<F>(inner: Arc<Inner>, mut callback: F)
+where
+    F: FnMut(SwitchPort, &'_ [u8]),
+{
+    // helper function to tag a socket's `.readable()` future with an index; we
+    // need this to make rustc happy about the types we push into
+    // `recv_all_sockets` below
+    async fn readable_with_port(
+        port: SwitchPort,
+        sock: &UdpSocket,
+    ) -> (SwitchPort, io::Result<()>) {
+        let result = sock.readable().await;
+        (port, result)
+    }
+
+    // set up collection of futures tracking readability of all switch port
+    // sockets
+    let mut recv_all_sockets = FuturesUnordered::new();
+    for (i, sock) in inner.ports.iter().enumerate() {
+        recv_all_sockets.push(readable_with_port(SwitchPort(i), sock));
+    }
+
+    let mut buf = [0; SpMessage::MAX_SIZE];
+
+    loop {
+        // `recv_all_sockets.next()` will never return `None` because we
+        // immediately push a new future into it every time we pull one out
+        // (to reregister readable interest in the corresponding socket)
+        let (port, result) = recv_all_sockets.next().await.unwrap();
+
+        // immediately push a new future requesting readability interest, as
+        // noted above
+        let sock = &inner.ports[port.0];
+        recv_all_sockets.push(readable_with_port(port, sock));
+
+        // TODO how do we handle errors here (or from `recv()` below)?
+        // currently just log them and retry, assuming any socket errors are
+        // ephemeral
+        if let Err(err) = result {
+            warn!(
+                inner.log,
+                "error checking socket readability";
+                "port" => ?port,
+                "err" => err,
+            );
+            continue;
+        }
+
+        match sock.try_recv_from(&mut buf) {
+            Ok((n, addr)) => {
+                if Some(addr) != inner.sp_socket(port).map(|s| s.addr) {
+                    // TODO what do we do here? we received a packet from an
+                    // address that doesn't match what we believe is the SP's
+                    // address. for now, log and discard
```

Comment:
this is a great point and we may want to use `TODO-security` as I've seen @davepacheco do. In particular, we might want to consider this situation as a trigger to potentially stop trusting a given sled or SP and force it to re-enter the trust quorum... or something.